### PR TITLE
Added antialias flag to interpolate (CPU only, bilinear)

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Added antialias flag to interpolate (CPU only)
  - forward and backward for bilinear mode
  - added tests

### Benchmarks

<details>
<summary>
Forward pass, CPU. PTH interpolation vs PIL
</summary>

Cases:
- PTH RGB 3 Channels, float32 vs PIL RGB uint8 (apples vs pears)
- PTH 1 Channel, float32 vs PIL 1 Channel Float

Code: https://gist.github.com/vfdev-5/b173761a567f2283b3c649c3c0574112

```
# OMP_NUM_THREADS=1 python bench_interp_aa_vs_pillow.py

Torch config: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_75,code=sm_75
  - CuDNN 8.0.5
  - Build settings: BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_PYTORCH_QNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=1, USE_CUDNN=1, USE_EIGEN_FOR_BLAS=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=OFF, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=0, USE_OPENMP=ON, 

Num threads: 1
[------------------------ Downsampling: torch.Size([1, 3, 906, 438]) -> (320, 196) ------------------------]
                                                  |  Reference, PIL 8.3.2, mode: RGB  |  1.10.0a0+git1e87d91
1 threads: -------------------------------------------------------------------------------------------------
      channels_first contiguous torch.float32     |                2.9                |          3.1        
      channels_last non-contiguous torch.float32  |                2.6                |          3.6        

Times are in milliseconds (ms).

[------------------------ Downsampling: torch.Size([1, 3, 906, 438]) -> (460, 220) ------------------------]
                                                  |  Reference, PIL 8.3.2, mode: RGB  |  1.10.0a0+git1e87d91
1 threads: -------------------------------------------------------------------------------------------------
      channels_first contiguous torch.float32     |                3.4                |          4.0        
      channels_last non-contiguous torch.float32  |                3.4                |          4.8        

Times are in milliseconds (ms).

[------------------------ Downsampling: torch.Size([1, 3, 906, 438]) -> (120, 96) -------------------------]
                                                  |  Reference, PIL 8.3.2, mode: RGB  |  1.10.0a0+git1e87d91
1 threads: -------------------------------------------------------------------------------------------------
      channels_first contiguous torch.float32     |                1.6                |          1.8        
      channels_last non-contiguous torch.float32  |                1.6                |          1.9        

Times are in milliseconds (ms).

[----------------------- Downsampling: torch.Size([1, 3, 906, 438]) -> (1200, 196) ------------------------]
                                                  |  Reference, PIL 8.3.2, mode: RGB  |  1.10.0a0+git1e87d91
1 threads: -------------------------------------------------------------------------------------------------
      channels_first contiguous torch.float32     |                9.0                |          11.3       
      channels_last non-contiguous torch.float32  |                8.9                |          12.5       

Times are in milliseconds (ms).

[----------------------- Downsampling: torch.Size([1, 3, 906, 438]) -> (120, 1200) ------------------------]
                                                  |  Reference, PIL 8.3.2, mode: RGB  |  1.10.0a0+git1e87d91
1 threads: -------------------------------------------------------------------------------------------------
      channels_first contiguous torch.float32     |                2.1                |          1.8        
      channels_last non-contiguous torch.float32  |                2.1                |          3.4        

Times are in milliseconds (ms).

[--------------- Downsampling: torch.Size([1, 1, 906, 438]) -> (320, 196) --------------]
                                 |  Reference, PIL 8.3.2, mode: F  |  1.10.0a0+git1e87d91
1 threads: ------------------------------------------------------------------------------
       contiguous torch.float32  |               1.2               |          1.0        

Times are in milliseconds (ms).

[--------------- Downsampling: torch.Size([1, 1, 906, 438]) -> (460, 220) --------------]
                                 |  Reference, PIL 8.3.2, mode: F  |  1.10.0a0+git1e87d91
1 threads: ------------------------------------------------------------------------------
       contiguous torch.float32  |               1.4               |          1.3        

Times are in milliseconds (ms).

[--------------- Downsampling: torch.Size([1, 1, 906, 438]) -> (120, 96) ---------------]
                                 |  Reference, PIL 8.3.2, mode: F  |  1.10.0a0+git1e87d91
1 threads: ------------------------------------------------------------------------------
       contiguous torch.float32  |              719.9              |         599.9       

Times are in microseconds (us).

[-------------- Downsampling: torch.Size([1, 1, 906, 438]) -> (1200, 196) --------------]
                                 |  Reference, PIL 8.3.2, mode: F  |  1.10.0a0+git1e87d91
1 threads: ------------------------------------------------------------------------------
       contiguous torch.float32  |               3.7               |          3.5        

Times are in milliseconds (ms).

[-------------- Downsampling: torch.Size([1, 1, 906, 438]) -> (120, 1200) --------------]
                                 |  Reference, PIL 8.3.2, mode: F  |  1.10.0a0+git1e87d91
1 threads: ------------------------------------------------------------------------------
       contiguous torch.float32  |              834.4              |         605.7       

Times are in microseconds (us).



```

</details>

Code is moved from torchvision: https://github.com/pytorch/vision/pull/3761, https://github.com/pytorch/vision/pull/3810 and https://github.com/pytorch/vision/pull/4208

### CODE REVIEW

**Review by fmassa — COMMENTED** [review]
I have some minor comments, otherwise good to me

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 1]
```diff
@@ -126,10 +126,63 @@ static inline scalar_t interpolate(char* src, char** data, const int64_t* stride
   return Interpolate<n, scalar_t, index_t, interp_size>::eval(src, data, strides, i);
 }
 
-template<int interp_size>
+template <typename scalar_t, typename index_t>
+static inline scalar_t interpolate_aa_single_dim_zero_strides(
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply aten/src/ATen/native/cpu/UpSampleKernel.cpp:130]: If we compare [step 2](https://github.com/vfdev-5/interpolate-antialiasing/blob/fd0cd49724c107dd7c216660a4cb2d8698eea2ed/step_two/aa_interpolation_impl.h#L104) vs [step 2.2](https://github.com/vfdev-5/interpolate-antialiasing/blob/fd0cd49724c107dd7c216660a4cb2d8698eea2ed/step_two_dot_two/aa_interpolation_impl.h#L132) the speed up should be the following:
```
Num threads: 1
[---------------------- Downsampling: torch.Size([3, 438, 906]) -> (320, 196) ----------------------]
                                 |  PIL SIMD 7.0.0.post3  |  1.9.0a0+gitb5647dd  |  aa_interp_lin_step_two
1 threads: ------------------------------------------------------------------------------------------
      channels_first contiguous  |       345.9       |        670.7         |          2927.0

Num threads: 1
[----------------------------------------------------- Downsampling: torch.Size([3, 438, 906]) -> (320, 196) -----------------------------------------------------]
                                 |  PIL 8.2.0  |  1.9.0a0+gitb5647dd  |  aa_interp_lin_step_two_dot_two 
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------
      channels_first contiguous  |    1842.8   |        664.0         |              2272.6

Times are in microseconds (us).
```

File: torch/nn/functional.py [hunk 2]
```diff
@@ -3728,6 +3736,8 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
         return torch._C._nn.upsample_linear1d(input, output_size, align_corners, scale_factors)
     if input.dim() == 4 and mode == "bilinear":
         assert align_corners is not None
+        if antialias:
+            return torch._C._nn.upsample_bilinear2d_aa(input, output_size, align_corners, scale_factors)
```
Comment (jbschlosser) [inline torch/nn/functional.py:3740]: Found a nice workaround to deal with the JIT FC window [here](https://github.com/pytorch/pytorch/issues/67565#issuecomment-961999987).

TL;DR: Gate the call to the new op appropriately using `torch.jit.is_scripting()` to avoid JIT FC issues while still allowing the new kwarg to be used outside of JIT scripting.

**Review by jbschlosser — COMMENTED** [review]
LGTM on the implementation side if @fmassa is good with it.

As usual, there are unfortunate JIT FC issues to be carefully dealt with - left a comment regarding this but g2g otherwise.

Also, to make future refactoring easier, I recommend making the new ops private as well by prepending with an underscore.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3833,6 +3841,12 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
         return torch._C._nn.upsample_linear1d(input, output_size, align_corners, scale_factors)
     if input.dim() == 4 and mode == "bilinear":
         assert align_corners is not None
+        # Enforce that the full call with the new kwarg is not invoked when scripting.
+        # TODO: Remove this scripting logic once the 2-week FC window has passed.
+        if antialias and torch.jit.is_scripting():
+            raise RuntimeError("TorchScript currently does not support antialias in interpolate")
+        if antialias:
```
Comment (jbschlosser) [inline torch/nn/functional.py:3848]: This branch needs to be gated by `not torch.jit.is_scripting()` as well. Note that scripting doesn't operate linearly - it will include all reachable code branches into the serialized result.

To achieve the goal of leaving out the call to `torch._C._nn._upsample_bilinear2d_aa` from the serialized JIT model, I'd suggest something like this:
```python
if antialias:
    if not torch.jit_scripting():
        return torch._C._nn._upsample_bilinear2d_aa(input, output_size, align_corners, scale_factors)
    else:
        raise RuntimeError("TorchScript currently does not support antialias in interpolate")
```

**Comment by jbschlosser** [comment]
Failing checks are related to ONNX (cc @BowenBao for assistance here)

**Comment by BowenBao** [comment]
error message `RuntimeError: 0INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/jit/ir/alias_analysis.cpp":606, please report a bug to PyTorch. We don't have an op for aten::__interpolate but it isn't a special case.  Argument types: Tensor, int, NoneType, str, NoneType, NoneType, bool,`, perhaps need to update signature, or add as special case? cc @eellison

**Comment by eellison** [comment]
@BowenBao this is an error with the https://github.com/pytorch/pytorch/blob/079573535193889abb0701e9d57b115554457ad8/torch/csrc/jit/passes/onnx/function_substitution.cpp pass. can you update the __interpolate node ? that is a really old workaround for one specific issue... is there any more general infrastructure we can use here ? I see the onnx has been working on exporting module calls

**Comment by BowenBao** [comment]
@eellison I see it's from here https://github.com/pytorch/pytorch/blob/079573535193889abb0701e9d57b115554457ad8/torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp#L797. Seems this is the workaround you mentioned. Curious do you know if the above `interpolate_op` and `interpolate` is also for workaround? They seem to be needing an update wrt the new `antialias` added in this PR too. Still need to BC support onnx without function export so I think need to keep and update these workarounds ..

**Comment by eellison** [comment]
`__interpolate__` isn't used anywhere in JIT, just in ONNX. so i think you can add the new arguments and add an error within the operator or whatever is needed for ONNX, i'm not sure.

**Comment by BowenBao** [comment]
Thank you @eellison .

@vfdev-5 You can unblock the CI issue by adding the new argument to https://github.com/pytorch/pytorch/blob/079573535193889abb0701e9d57b115554457ad8/torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp#L797 and https://github.com/pytorch/pytorch/blob/079573535193889abb0701e9d57b115554457ad8/torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp#L700 and throw error if `antialias` is set to True. This should unblock this PR. We will follow-up with ONNX support if needed.

**Comment by vfdev-5** [comment]
@BowenBao thanks for the suggestions. I updated locally the code and running the test:
```
test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset10::test_interpolate_downsample
```
If I update only `aten::__interpolate` and not others like `aten::__interpolate.scale_list`, ...
then I have the same issue as here:
```

```
If I update all methods then I have below issue:
```
  File "/pytorch/torch/onnx/utils.py", line 1084, in _run_symbolic_function
    return symbolic_fn(g, *inputs, **attrs)
TypeError: __interpolate() takes 7 positional arguments but 8 were given
```

Any suggestions on how to fix that ? Thanks !

**Comment by BowenBao** [comment]
https://github.com/pytorch/pytorch/blob/746a31b2909d99bada381da4deec86c36a6083b6/torch/onnx/symbolic_opset9.py#L1090
Please add the new argument there as well. For symbolic_opset8.py, symbolic_opset9.py, symbolic_opset10.py and symbolic_opset11.py

**Review by jbschlosser — APPROVED** [review]
LGTM thanks :)

File: aten/src/ATen/native/cpu/UpSampleKernel.cpp [hunk 4]
```diff
@@ -1002,6 +1363,135 @@ void upsample_nearest3d_backward_kernel_impl(
...
+
+  auto grad_output_data = grad_output.data_ptr<scalar_t>();
+  auto grad_input_data = grad_input.data_ptr<scalar_t>();
+  auto input_sizes = grad_input.sizes().vec();
+  auto output_sizes = grad_output.sizes().vec();
+  auto ndim = input_sizes.size();
+
+  // treat nbatch and channels as one dimension
+  int64_t channels = input_sizes[0] * input_sizes[1];
+  int64_t input_depth = (ndim == 5) ? input_sizes[2] : 1;
```
Comment (jbschlosser) [inline aten/src/ATen/native/cpu/UpSampleKernel.cpp:1389]: @vfdev-5 FYI internal linter checks are failing because this variable is unused. Unfortunately they're a tad more strict that the ones visible publicly

**Comment by cpuhrsch** [comment]
@jbschlosser - this didn't update InterpolateFuncOptions of the C++ API.

**Comment by vfdev-5** [comment]
@cpuhrsch thanks for noting. I can update it in a follow-up PR. 
Is it also applicable for "nearest-exact" mode introduced in https://github.com/pytorch/pytorch/pull/64501 ?

**Comment by cpuhrsch** [comment]
@vfdev-5 - looks like it

**Comment by jbschlosser** [comment]
Thanks - it's unfortunate that the C++ API has a wholly separate implementation of `interpolate()`.
